### ISSUE TITLE
Apollo didn't turn right, maybe a bug

### ISSUE THREAD

**Issue by 1Jyfc** [ApolloAuto/apollo#14842]
### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 18.04)**: Linux Ubuntu 18.04
- **Apollo installed from (source or binary)**: source
- **Apollo version (3.5, 5.0, 5.5, 6.0)**: 6.0
- **Output of `apollo.sh config` if on `master` branch**:

### Steps to reproduce the issue:
Hello Apollo team! Recently I am testing Apollo 6.0 with LGSVL and I found something that may be a bug of Planning module. Here is the steps of testing:

- The case is in map BorregasAve, as this picture shows:
![Screenshot from 2023-03-22 22-23-42](https://user-images.githubusercontent.com/32264630/226936106-db5a11f0-9b83-455a-ad27-a80eb2393dae.png)
I use LGSVL as simulator of Apollo. For ego configuration, I use 3D ground truth and signal sensor to replace Perception module, which ensure the correctness of NPC cars and signals.

- The destination point is on the road right of junction. In this case, the ego should deal with scenario TRAFFIC LIGHT UNPROTECTED RIGHT TURN. And since I did not turn on config `enable_right_turn_on_red` in `modules/planning/conf/scenario/traffic_light_unprotected_right_turn_config.pb.txt`, the ego will stop before the stop line of junction. At this moment, in `cyber_monitor`, topic of traffic light is here:
![Screenshot from 2023-03-22 22-23-45](https://user-images.githubusercontent.com/32264630/226937863-5a7a0af2-9a3c-4257-85b6-0c806e83afaf.png)
4 items represent 4 signals in picture. In `cyber_monitor` they are all red now.

- After starting simulation, ego will drive to stop line and then stop:
![Screenshot from 2023-03-22 22-24-05](https://user-images.githubusercontent.com/32264630/226938398-1612b501-8be9-4810-9ee0-1f53115cfe7f.png)
in picture we just see 3 signals, and the last signal is here:
![Screenshot from 2023-03-22 22-24-18](https://user-images.githubusercontent.com/32264630/226938828-255405e4-fc1c-435f-a3f6-f1e4f0ab3b4c.png)
At this moment, `cyber_monitor` shows:
![Screenshot from 2023-03-22 22-24-08](https://user-images.githubusercontent.com/32264630/226939279-9403a745-7b39-426c-b400-f4e07a47cb35.png)
Seems that the last signal cannot be seen by ego, which is reasonable.

- However, after signals turn green, ego did not run, like this:
![Screenshot from 2023-03-22 22-24-40](https://user-images.githubusercontent.com/32264630/226940565-c2f11e2b-b85c-4a5e-87c2-8aefd9d6a70f.png)
Based on log files, I found that the stage of Planning module at this moment is STOP in TRAFFIC LIGHT UNPROTECTED RIGHT TURN. Here is the code:
```
for (const auto& traffic_light_overlap_id :
       GetContext()->current_traffic_light_overlap_ids) {
    // get overlap along reference line
    current_traffic_light_overlap = scenario::util::GetOverlapOnReferenceLine(
        reference_line_info, traffic_light_overlap_id,
        ReferenceLineInfo::SIGNAL);
    if (!current_traffic_light_overlap) {
      continue;
    }

    // set right_of_way_status
    reference_line_info.SetJunctionRightOfWay(
        current_traffic_light_overlap->start_s, false);

    const double distance_adc_to_stop_line =
        current_traffic_light_overlap->start_s - adc_front_edge_s;
    auto signal_color = frame->GetSignal(traffic_light_overlap_id).color();
    ADEBUG << "traffic_light_overlap_id[" << traffic_light_overlap_id
           << "] start_s[" << current_traffic_light_overlap->start_s
           << "] distance_adc_to_stop_line[" << distance_adc_to_stop_line
           << "] color[" << signal_color << "]";

    // check distance to stop line
    if (distance_adc_to_stop_line >
        scenario_config_.max_valid_stop_distance()) {
      traffic_light_all_stop = false;
      break;
    }

    // check on traffic light color
    if (signal_color != TrafficLight::GREEN) {
      traffic_light_all_green = false;
      traffic_light_no_right_turn_on_red =
          CheckTrafficLightNoRightTurnOnRed(traffic_light_overlap_id);
      break;
    }
  }

  if (traffic_light_all_stop && traffic_light_all_green) {
    return FinishStage(true);
  }

  if (!traffic_light_no_right_turn_on_red) {
    if (traffic_light_all_stop && !traffic_light_all_green) {
      // check distance pass stop line
      const double distance_adc_pass_stop_line =
          adc_front_edge_s - current_traffic_light_overlap->end_s;
      ADEBUG << "distance_adc_pass_stop_line[" << distance_adc_pass_stop_line
             << "]";
      if (distance_adc_pass_stop_line >
          scenario_config_.min_pass_s_distance()) {
        return FinishStage(false);
      }

      if (scenario_config_.enable_right_turn_on_red()) {
        // when right_turn_on_red is enabled
        // check on wait-time
        if (GetContext()->stop_start_time == 0.0) {
          GetContext()->stop_start_time = Clock::NowInSeconds();
        } else {
          auto start_time = GetContext()->stop_start_time;
          const double wait_time = Clock::NowInSeconds() - start_time;
          ADEBUG << "stop_start_time[" << start_time << "] wait_time["
                 << wait_time << "]";
          if (wait_time >
              scenario_config_.red_light_right_turn_stop_duration_sec()) {
            return FinishStage(false);
          }
        }
      }
    }
  }
```
For the signal part, seems that Apollo will check all the traffic light overlap, if all of them are green, `traffic_light_all_stop` will be set as true, which causes stage transition. Besides, due to logs added by my codes, I found that a traffic light overlap seems to be created when simulation starts. That is, at the first time ego can see the signal beside the road, so ego creates a traffic light overlap for it. However, when ego passes the signal, it cannot see it any more, but the overlap is still in `GetOverlapOnReferenceLine`. So it will still be checked by Planning module, but the color of signal is always `UNKNOWN`, which stops the ego permanently.

The same problem will happen in LEFT TURN scenario too. I do not know whether it is a bug, and I cannot find related issues either. Hope for your reply, thanks!

**Comment by YuqiHuai** [comment]
Hi @1Jyfc 
Can you share some screenshots of Dreamview during the scenario? It would be extremely helpful to share a cyber record file recorded via `cyber_recorder record -a`

I am aware of a similar issue that cause the planning module to fail moving forward, https://github.com/ApolloAuto/apollo/issues/14551 causes Apollo to stop indefinitely on top of a stop line. But to trigger this bug, Apollo must be initialized directly on top of a stop line. Based on your screenshot, you initialized Apollo a few meters in front of the stop line, therefore it should be a different issue.

Once you share the cyber record file, it will be much easier for others to look at the details of the planning messages.

**Comment by qwetqwe** [comment]
You can see dreamview/PNC_Monitor or planning topic to find the reason why vehicle stop

**Comment by 1Jyfc** [comment]
Thanks for your reply! 
[20230323162844.record.zip](https://github.com/ApolloAuto/apollo/files/11048926/20230323162844.record.zip) @YuqiHuai Here is the cyber record file. I cannot use python API to record, because it will cause this error:
`FileNotFoundError: [Errno 2] No such file or directory: "docker exec -u $USER apollo_dev_jyf pkill -f 'cyber_recorder record'"`
so I record by shell.

And I add some log codes in planning module to log out some useful information, mainly in `modules/planning/scenarios/traffic_light/unprotected_right_turn/stage_stop.cc`. And according to [this log](https://github.com/ApolloAuto/apollo/files/11049043/planning.log.INFO.20230323-170051.zip), stage of planning module switches into right turn after 7 check points, and at first time the status is:
![Screenshot from 2023-03-23 17-15-53](https://user-images.githubusercontent.com/32264630/227157153-a9ae917f-8cd1-48ba-9953-451a57686a1a.png)

At the first check point of scenario "right turn" stage "stop", 4 overlaps are created for 4 signals, and at this time they are all red. I tried to figure out the SL position of overlap and ego and printed out. I found that the 4 overlaps share the same S.

And after several check points, the ego passes the fourth signal, but doesn't pass the overlap, so it is still checked, and its color changes into UNKNOWN, both in log and Dreamview (Here, when I replay the record, in Dreamview the traffic light always remain red, but when I use LGSVL it will show UNKNOWN). After other signals turn green, it still remains UNKNOWN, which causes `traffic_light_all_stop && traffic_light_all_green` cannot be satisfied. Here `!traffic_light_no_right_turn_on_red && traffic_light_all_stop && !traffic_light_all_green` is satisfied, but since ego doesn't pass stop line, `distance_adc_pass_stop_line > scenario_config_.min_pass_s_distance()` cannot be satisfied. So in this case if configuration of right turn on red light is disabled, the ego will not drive any more. In fact at this time ego is still ahead of stop line, so the overlap will not be canceled.

I think this case is like #14551 but not totally the same. In my log I also find the same logs, but if the traffic light can be considered as GREEN, it will cause stage switch.

**Comment by YuqiHuai** [comment]
I spent some more time thinking about this scenario and I agree with this being a bug in Apollo. But the fix might be complicated. @1Jyfc @qwetqwe 

<img width="600" alt="image" src="https://user-images.githubusercontent.com/34195403/227426153-9c232c30-ce13-49c8-b7d1-e961eb008d80.png">

Above is an illustration of signals and the stop line the signal is controlling at the junction. 

![image](https://user-images.githubusercontent.com/34195403/227426648-0b11b89e-f58c-49da-97d9-2d981d0b9166.png)


Above is the screenshot of Apollo at the junction in the SVL simulator. Boxed in red is `signal_9`.

---

In a realistic environment with Apollo's perception, I do not think Apollo could detect `signal_9` in any way because it has already moved passed it. Therefore theoretically at this exact junction, Apollo can only make a right turn in simulations (and someone has to correct the TrafficLightDetection to include `signal_9`); in the real world, Apollo won't detect `signal_9` and will stop indefinitely at this stop line.

The root cause of the bug should be Apollo looping through `GetContext()->current_traffic_light_overlap_ids` and checking every signal that overlaps, ignoring the fact that at certain locations, not all signals can be detected.

**Comment by YuqiHuai** [comment]
I have attempted to provide a fix to address the issue. 

At a high level: 1) stop stage is only interested in overlapping signals that are detected; if an overlapping signal is not detected, consider it done.

# Discussion

https://github.com/YuqiHuai/apollo/blob/3ee79789c2bbcf470b9f9a4e3cd345f4451afdee/modules/planning/scenarios/traffic_light/unprotected_right_turn/stage_stop.cc#L67-L85

https://github.com/YuqiHuai/apollo/blob/3ee79789c2bbcf470b9f9a4e3cd345f4451afdee/modules/planning/scenarios/traffic_light/unprotected_right_turn/stage_stop.cc#L104

This part of the code maps to high-level goal 1. This part ensures planning can transition from STOP to CRUISE stage. However, this is not enough to fix the issue. The stop fence for `signal_9` is still not removed, planning's main decision is still stopping for `signal_9`. Shown in the image (left) is a much lighter stop fence, compared to the unfixed version (right) with a much darker stop fence (probably because many are stacked together)

| with partial fix | original |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/34195403/227450110-549eed12-3ddb-493a-8a4c-2e4474c1e4dc.png) | ![image](https://user-images.githubusercontent.com/34195403/227452288-b4774027-ff88-468d-b44f-c0a2eab5c035.png) |

---

I did not find the most appropriate place to address the above issue therefore I thought of high-level goal 2. 

https://github.com/YuqiHuai/apollo/blob/3ee79789c2bbcf470b9f9a4e3cd345f4451afdee/modules/planning/scenarios/traffic_light/unprotected_right_turn/stage_stop.cc#L87-L101

This considers overlapping signals that are not in interested signals as "done", so signal_9 is considered done once it is no longer detected, its stop fence will also be removed.

---

![image](https://user-images.githubusercontent.com/34195403/227451342-ca6f6f8d-294b-437e-ba52-3130533f22e8.png)

Using the temporary fix I provided, observe the bug symptom is gone. Notice in the screenshot I am replaying `reproduce.record` which originally will lead to a scenario in which Apollo stops indefinitely at this stop line.

Any feedback is highly appreciated and thanks @1Jyfc for finding this interesting scenario!

**Comment by 1Jyfc** [comment]
@YuqiHuai Thanks for your feedback! I just discuss this issue with some experts and I also consider it as a bug. For the question about whether signal sensor can provide true messages, I tried not to use signal sensor, but to use perception module of Apollo, which got the same result. And in real world, in this situation I do not think ego can detect such signal either.

In fact, in log file we can find that the 4 overlaps created for 4 signals share the same area of reference line, which represents the stop line. And when a signal is passed by ego, it cannot be detected any more, but its overlap is still in scenario, which will be repeatedly checked.

I think the logic of considering passed light is meaningful, which avoid the situation that ego passes the red light too fast. But in this case 4 signals share the same overlap, then I think the judgment about signals related to this overlap should ignore the unseen signal with the same overlap.

**Comment by YuqiHuai** [comment]
@1Jyfc Sounds good. This is a great scenario revealing correct perception output may not be what the planning module expects as the "correct input". In this case, planning is expecting 4 TL detections while perception can only realistically detect 3.

The analysis should help @qwetqwe to evaluate and confirm this being a bug, and my temporary fix should be able to justify the root cause of over consideration (i.e., considering a traffic light that is not detectable). This bug should also impact all other traffic light related decision making. You can verify this by sending a routing request that asks the ego car to go straight instead of turning right in the scenario you reported, I imagine Apollo will also stop indefinitely due to not being able to detect `signal_9`.

You are saying "considering passed light is meaningful", can you elaborate a bit more on what kind of scenario may cause issues related to red lights? I disagree with this a bit because intuitively, a human driver would not make decisions based on a traffic light that he cannot see; if he cannot see any lights, then it is safer to stop at the stop line. Therefore a fix matching this intuition is: only consider detected lights and make a conservative if 1) no light is detected or 2) detected lights for the current overlap have different colors.

**Comment by 1Jyfc** [comment]
> You are saying "considering passed light is meaningful", can you elaborate a bit more on what kind of scenario may cause issues related to red lights? I disagree with this a bit because intuitively, a human driver would not make decisions based on a traffic light that he cannot see; if he cannot see any lights, then it is safer to stop at the stop line. Therefore a fix matching this intuition is: only consider detected lights and make a conservative if 1) no light is detected or 2) detected lights for the current overlap have different colors.

@YuqiHuai I considered this problem again and I think you are right. In real world, drivers should not do decisions based on unseen lights. The final problem is to deal with the conflict that ego passes the light but doesn't pass the overlap. Thanks for your feedback!

**Comment by YuqiHuai** [comment]
@1Jyfc Sounds good. Thank you for your feedback as well! This has been a long but interesting discussion.

**Comment by LordonCN** [comment]
@YuqiHuai  Thanks for your helpfu reply

**Comment by YuqiHuai** [comment]
@LordonCN Hi! thanks for your feedback!

It is a bit rare to see an issue getting closed without either confirming this to be a bug or actually fixing the issue. It is pretty clear this should be a bug and I am not 100% if the fix I suggested is acceptable to the core developers.

Will any of the Apollo members (e.g., qwetqwe, daohu527) be looking at this more in the future?

**Comment by LordonCN** [comment]
> @LordonCN Hi! thanks for your feedback!
> 
> It is a bit rare to see an issue getting closed without either confirming this to be a bug or actually fixing the issue. It is pretty clear this should be a bug and I am not 100% if the fix I suggested is acceptable to the core developers.
> 
> Will any of the Apollo members (e.g., qwetqwe, daohu527) be looking at this more in the future?

Hello YuqiHuai
I have already told him(@qwetqwe) about this and we appreciate it if you could create a pull request.

**Comment by YuqiHuai** [comment]
@LordonCN Oh! That sounds good. Thanks for responding! I'll create a pull request for the right-turn scenario fix and continue adding the fix to other scenarios to the same PR.

**Comment by qwetqwe** [comment]
Thanks for attention @YuqiHuai @1Jyfc . The main reason is that planning did not get signal_9 status. I think it is map(signal stop line should be behind signal boundary) or lgsvl(simulator should send all traffic lights in front of vehicle path) problem.

**Comment by YuqiHuai** [comment]
Hi @qwetqwe thanks for the feedback! 

This might not be an LGSVL problem because its ground truth seems to be modeling ADS's perception module's field of view correctly.

Will this also be considered a planning problem as the planning module is over-expecting information that cannot be detected? Maybe the signal's stop line can be positioned differently, but there are other situations that may cause certain signals to be undetectable (i.e., an attacker purposely blocks a traffic signal).

In summary, is it necessary for planning to see every signal controlling a stop line based on the HD Map, or is it enough to only consider signals that have been detected?

**Comment by qwetqwe** [comment]
@YuqiHuai Apollo is only a template for developers' secondary development. It can not cover too many corner cases.That will make it hard to learn and maintain.

**Comment by YuqiHuai** [comment]
@qwetqwe Thanks for the explanation! I think I get your point, if specific logic is implemented to handle each corner case, then the documentation is probably no longer manageable. I have a quick follow-up question:

Is it correct, that issues and PRs that discover and handle difficult-to-find situations (like this one) will generally not be considered bugs, and fixes will not be merged?
